chichiwang/exercise-js
# Use `project-configs.js` to define server port based on build environment

## Description
Linked to Issue: #125
Create a configuration file for the project `project-configs.js` which is git-ignored (configurations can be project-specific and should not be tracked in the case they either contain business logic to retrieve or secrets).

A sample file to show the format, `project-configs-example.js` is added to the project to serve as example for the structure expected of the configurations. A npm script `create-config` is added to copy `project-configs-example.js` to a new file `project-configs.js` only if that file does not already exist.

These configurations are fed into the build step via [Webpack DefinePlugin](https://webpack.js.org/plugins/define-plugin/) and gated against production vs non-production build environments: the `development` configs will be used for non-production builds, and the `production` configs used for production builds.

The first config value, `PORT` is added to set the server port to `8080` for non-production builds and `3000` for production builds. `server/index.ts` is updated to use this configuration via the Webpack global `PORT`.

Documentation is updated in both `docs/README.md` and `pages/Documentation/index.tsx` to:
* Add a step to run `npm run create-config` in the section "Starting a Project"
* Add documentation for the script `create-config` in the section "NPM Scripts"
* Add a section under "Developing Locally" called "Project Configurations" to outline how project configurations work, and how to add one

## Changes
* Add "project-configs.js" to `.gitignore`
* Increase `p` tag `line-height` to `1.75rem` in `app/global.module.scss`
* Declare a constant `PORT` in `app/globals.d.ts`
* Update `docs/README.md`
  * Add step to section: Starting a Project
  * Add documentation for script `create-config` in section: NPM Scripts
  * Add section: Project Configurations
* Add script `create-config` to `package.json`
  * Copy `project-configs-example.js` to a new file `project-configs.js` if it does not already exist
* Update `pages/Documentation/NPMScripts.tsx` to add documentation for script `create-config`
* Create a `pages/Documentation/ProjectConfigurations.tsx` to document project configuration
  * Mirrors content in `docs/README.md` in the section: Project Configurations
* Update `pages/Documentation/StartingAProject.tsx`
  * Add a step to run `npm run create-config`
* Update `pages/Documentation/index.tsx`
  * Render section: Project Configurations
  * Create link in "Developing Locally" to link to section: Project Configurations
* Add a `project-configs-example.js` file with a sample project config
* Update `server/index.ts` to read the port from the project config variable `PORT`
  * Update tests
* Create a `webpack/define.js` module to feed project config into the Webpack DefinePlugin based on build environment
  * Plug this plugin instance into `webpack/client.config.js` and `webpack/server.config.js`

## Steps to QA
* Pull down and switch to this branch
* Without generating a `project-configs.js` file:
  * Run `npm run watch`, `npm run dev`, `npm run prod`, or any of the build scripts to ensure they fail with a meaningful error message
  * Run `npm run create-config` to ensure that a `project-configs.js` file is generated and its contents match `project-configs-example.js`
* After having generated a `project-configs.js` file:
  *  Make a change to the file
    * Run `npm run create-config` and ensure the contents of that file have not changed
    * Ensure the change is not picked up by git (`git status`)
* Run the app in development mode (`npm run watch`/`npm run dev`) and ensure the server is listening on the port specified for development in `project-config.js`
* Run the app in production mode (`npm run prod`) and ensure the server is listening on the port specified for production in `project-config.js`
* Review the changes in `docs/README.md` to make sure they read well
* Run the app and review the changes on the `/documentation` page to ensure they look correct
  * Check the contents of the copy button in the sections "NPM Scripts" and "Project Configurations" to ensure they are correct
